### PR TITLE
closes #593 implemented async commit

### LINKED ISSUE #593: Move commit processing to other threads
After working on #592 I am thinking that all commit processing could be pushed off to other threads.  Currently the threads processing transactions also do the commit.  Doing the commit involves waiting on results to come back from the conditional writer multiple times.  I am thinking that instead of doing this threads should push transactions that are ready to commit on to a queue where a few threads process them.  This queue would be memory limited.   This would allow more work to be built up for the conditional writers and batch writers (as much as will fit in memory).  As soon as a thread is ready to commit a tx, it would throw that on a queue and immediately go start working on getting another transaction ready to commit.  I have a strong suspicion that this could really increase throughput.

### PR DESCRIPTION
**Pull request by keith-turner**

### CODE REVIEW

File: modules/core/src/main/java/io/fluo/core/async/AsyncTransaction.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+package io.fluo.core.async;
+
+import io.fluo.api.client.Transaction;
+import io.fluo.core.impl.TxStats;
+
+public interface AsyncTransaction extends Transaction {
+  public TxStats getStats();
+
+  public int getSize();
```
Comment (mikewalch) [inline modules/core/src/main/java/io/fluo/core/async/AsyncTransaction.java]: `public` is redundant for interface methods.

File: modules/core/src/main/java/io/fluo/core/impl/TransactionImpl.java [hunk 2]
```diff
@@ -411,164 +427,31 @@ public boolean preCommit(CommitData cd) throws TableNotFoundException, AccumuloE
 
   }
 
+  @VisibleForTesting
   public boolean preCommit(CommitData cd, Bytes primRow, Column primCol)
```
Comment (mikewalch) [inline modules/core/src/main/java/io/fluo/core/impl/TransactionImpl.java]: `primRow` and `primCol` are never used

File: modules/core/src/main/java/io/fluo/core/impl/TransactionImpl.java [hunk 3]
```diff
@@ -411,164 +427,31 @@ public boolean preCommit(CommitData cd) throws TableNotFoundException, AccumuloE
...
 
-    if (cd.getRejected().size() > 0) {
-      rollback(cd);
-
-      if (ackCollision) {
-        throw new AlreadyAcknowledgedException();
-      }
-
+    SyncCommitObserver sco = new SyncCommitObserver();
+    beginCommitAsync(cd, sco);
```
Comment (mikewalch) [inline modules/core/src/main/java/io/fluo/core/impl/TransactionImpl.java]: could this be called `preCommitAsync`

File: modules/core/src/main/java/io/fluo/core/impl/TransactionImpl.java [hunk 4]
```diff
@@ -656,302 +540,597 @@ private boolean checkForAckCollision(ConditionalMutation cm) {
...
+        deleteLocks(cd, commitTs);
+      }
+    }
   }
 
-  @Override
-  public long getStartTimestamp() {
-    return startTs;
-  }
+  public void deleteLocks(CommitData cd, final long commitTs) {
```
Comment (mikewalch) [inline modules/core/src/main/java/io/fluo/core/impl/TransactionImpl.java]: can be private

File: modules/core/src/main/java/io/fluo/core/impl/TransactionImpl.java [hunk 5]
```diff
@@ -656,302 +540,597 @@ private boolean checkForAckCollision(ConditionalMutation cm) {
...
+  public boolean finishCommit(CommitData cd, Stamp commitStamp) throws TableNotFoundException,
+      MutationsRejectedException {
+    deleteLocks(cd, commitStamp.getTxTimestamp());
+    return true;
   }
 
-  @Override
-  public Map<String, Map<Column, String>> gets(Collection<RowColumn> rowColumns) {
-    return TxStringUtil.gets(this, rowColumns);
+  public void finishCommit(CommitData cd, long commitTs) {
```
Comment (mikewalch) [inline modules/core/src/main/java/io/fluo/core/impl/TransactionImpl.java]: can be private

File: modules/core/src/main/java/io/fluo/core/impl/TransactionImpl.java [hunk 6]
```diff
@@ -656,302 +540,597 @@ private boolean checkForAckCollision(ConditionalMutation cm) {
...
+      protected void onSuccess(CommitData cd, Iterator<Result> result) throws Exception {
+        postCommitPrimary(cd, commitTs, delLockMutation, Iterators.getOnlyElement(result));
+      }
+    }, env.getSharedResources().getCommitExecutor());
   }
 
-  private synchronized void checkIfOpen() {
-    if (status != TxStatus.OPEN) {
-      throw new IllegalStateException("Transaction is no longer open! status = " + status);
+  protected void postCommitPrimary(CommitData cd, long commitTs,
```
Comment (mikewalch) [inline modules/core/src/main/java/io/fluo/core/impl/TransactionImpl.java]: can be private

File: modules/core/src/main/java/io/fluo/core/impl/TransactionImpl.java [hunk 7]
```diff
@@ -656,302 +540,597 @@ private boolean checkForAckCollision(ConditionalMutation cm) {
...
+        commmitPrimary(cd, commitTs);
+      }
+    }, env.getSharedResources().getCommitExecutor());
+
   }
 
-  @Override
-  public void close() {
-    close(true);
+  protected void commmitPrimary(CommitData cd, final long commitTs) {
```
Comment (mikewalch) [inline modules/core/src/main/java/io/fluo/core/impl/TransactionImpl.java]: can be private

File: modules/core/src/main/java/io/fluo/core/impl/TransactionImpl.java [hunk 8]
```diff
@@ -656,302 +540,597 @@ private boolean checkForAckCollision(ConditionalMutation cm) {
...
 
-  /**
-   * Retrieves transactor ID by first getting/creating transactor (which is only done until
-   * necessary)
-   */
-  private Long getTransactorID() {
-    if (tnode == null) {
-      tnode = env.getSharedResources().getTransactorNode();
+
+  public void beginSecondCommitPhase(CommitData cd, Stamp commitStamp) throws Exception {
```
Comment (mikewalch) [inline modules/core/src/main/java/io/fluo/core/impl/TransactionImpl.java]: can be private

File: modules/core/src/main/java/io/fluo/core/impl/TransactionImpl.java [hunk 9]
```diff
@@ -391,6 +401,12 @@ private void addToRejected(Bytes row, Set<Column> columns) {
     public String toString() {
       return prow + " " + pcol + " " + pval + " " + rejected.size();
     }
+
+    // async stuff
+    private AsyncConditionalWriter acw;
+    public AsyncConditionalWriter bacw;
```
Comment (mikewalch) [inline modules/core/src/main/java/io/fluo/core/impl/TransactionImpl.java]: can be package local

File: modules/core/src/main/java/io/fluo/core/impl/SharedBatchWriter.java [hunk 10]
```diff
@@ -160,31 +160,42 @@ public void writeMutations(Collection<Mutation> ml) {
     }
   }
 
-  public void close() {
+  private static final Runnable DO_NOTHING = new Runnable() {
+    @Override
+    public void run() {}
+  };
+
+  public ListenableFuture<Void> writeMutationsAsyncFuture(Collection<Mutation> ml) {
```
Comment (mikewalch) [inline modules/core/src/main/java/io/fluo/core/impl/SharedBatchWriter.java]: Several methods in this class do not need to be public and can be package local

File: modules/cluster/src/main/java/io/fluo/cluster/runner/AppRunner.java [hunk 11]
```diff
@@ -323,12 +320,7 @@ public long countNotifications(Environment env) {
 
     Notification.configureScanner(scanner);
 
-    long count = 0;
-    for (Iterator<Map.Entry<Key, Value>> iterator = scanner.iterator(); iterator.hasNext(); iterator
-        .next()) {
-      count++;
-    }
-    return count;
+    return Iterables.size(scanner);
```
Comment (mikewalch) [inline modules/cluster/src/main/java/io/fluo/cluster/runner/AppRunner.java]: nice

File: modules/core/src/main/java/io/fluo/core/async/AsyncCommitObserver.java [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+ * 
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.fluo.core.async;
+
+public interface AsyncCommitObserver {
```
Comment (mikewalch) [inline modules/core/src/main/java/io/fluo/core/async/AsyncCommitObserver.java]: Some documentation for this interface would be nice.

File: modules/core/src/main/java/io/fluo/core/client/FluoClientImpl.java [hunk 13]
```diff
@@ -66,7 +66,7 @@ public FluoClientImpl(FluoConfiguration config) {
   @Override
   public LoaderExecutor newLoaderExecutor() {
     try {
-      return new LoaderExecutorImpl(config, env);
+      return new LoaderExecutorAsyncImpl(config, env);
```
Comment (mikewalch) [inline modules/core/src/main/java/io/fluo/core/client/FluoClientImpl.java]: Is there a reason for adding Async to class name?

**Comment by mikewalch** [comment]
+1

File: modules/core/src/main/java/io/fluo/core/impl/TransactionImpl.java (hunk 3 above)
Comment (keith-turner) [inline modules/core/src/main/java/io/fluo/core/impl/TransactionImpl.java]: no, it starts the entire async commit process...

File: modules/core/src/main/java/io/fluo/core/impl/TransactionImpl.java (hunk 2 above)
Comment (keith-turner) [inline modules/core/src/main/java/io/fluo/core/impl/TransactionImpl.java]: nice find

File: modules/cluster/src/main/java/io/fluo/cluster/runner/AppRunner.java (hunk 11 above)
Comment (keith-turner) [inline modules/cluster/src/main/java/io/fluo/cluster/runner/AppRunner.java]: yeah thats a little jewel that @ericnewton  showed me...

File: modules/core/src/main/java/io/fluo/core/client/FluoClientImpl.java (hunk 13 above)
Comment (keith-turner) [inline modules/core/src/main/java/io/fluo/core/client/FluoClientImpl.java]: not really... while working on this I had two implementations for a bit (the old sync one and the new async one)... then I deleted the old sync classes

**Pull request merged**